modelhub/modelhub/decorators.py

```diff
@@ -0,0 +1,123 @@
+import inspect
+from functools import wraps
+from typing import List, Union, Optional, TYPE_CHECKING
+from typing_extensions import Protocol
+
+import bach
+from sql_models.constants import not_set
+
+
+if TYPE_CHECKING:
+    from modelhub import Map, Aggregate
+
+
+class ModelFunctionType(Protocol):
+    @property
+    def __name__(self) -> str:
+        ...
+
+    def __call__(
+        self, _self: Union['Map', 'Aggregate'], data: bach.DataFrame, *args, **kwargs
+    ) -> bach.Series:
+        ...
+
+
+def use_only_required_objectiv_series(
+    required_series: Optional[List[str]] = None,
+    include_series_from_params: Optional[List[str]] = None,
+):
+    """
+    Internal: Decorator for validating and limiting the series used on a function dedicated
+    to generate new aggregated series based on supported objectiv columns.
+
+    :param required_series: A list of objectiv series names that the DataFrame must have
+    :param include_series_from_params: A list of parameters containing series names to be considered
+    in the dataframe.
+
+    The main purposes of the decorator are to:
+        * Validate that the dataframe passed to the function contains all required series for the
+        calculation,
+        * Optimize the base node of the returned series, as materialization is always applied after
+        aggregations. If the passed dataframe includes unused series containing complex expressions,
+        such expressions will remain on the final base node's history, meaning that the final query
+        might contain extra information that is not related to the expected result. Currently, bach
+        does not perform any optimization over this scenario, therefore ModelHub must be in charge
+        of ensuring the final query contains only the columns needed for all calculations.
+    """
+    def check_objectiv_data_decorator(func: ModelFunctionType):
+        @wraps(func)
+        def wrapped_function(
+            _self: Union['Map', 'Aggregate'],
+            data: bach.DataFrame,
+            *args,
+            **kwargs,
+        ) -> bach.Series:
+            from modelhub.stack.util import check_objectiv_dataframe
+            check_objectiv_dataframe(
+                df=data,
+                columns_to_check=required_series,
+                check_index=True,
+                check_dtypes=True,
+                with_md_dtypes=True,
+            )
+
+            extra_series = _get_extra_series_to_include_from_params(
+                func, data, include_series_from_params, *args, **kwargs,
+            )
+            series_to_include = list(set(required_series or {}) | set(extra_series))
```

Comment:
Didn't test this, but I think this should be
```suggestion
            series_to_include = set(required_series or {}) | set(extra_series)
```
The next line is `data = data[series_to_include]`, and in `DataFrame.__getitem__()` we use the order of the columns in a list, but if we pass a set we respect the original order. I think it would be weird if this decorator changes the order of the columns